apps/studio/src/lib/menu/tableMenu.ts

```diff
@@ -130,6 +130,53 @@ export async function copyRange(options: {
   options.range.getElement().classList.add("copied");
 }
 
+export function pasteRange(range: Tabulator.RangeComponent) {
+  const text = ElectronPlugin.clipboard.readText();
+  if (!text) return;
+
+  const parsedText = Papa.parse(text, {
+    header: false,
+    delimiter: '\t',
+  });
+
+  if (parsedText.errors.length > 0) {
+    const cell = range.getCells()[0]
+    setCellValue(cell, text);
+  } else {
+    const table = range.getCells()[0].getTable()
+    const rows = table.modules.spreadsheet.getRows().slice(range.getTop())
+    /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+    // @ts-ignore
+    // TODO: use range.getLeft(), which is not available in tabulator yet
+    const columns = table.modules.spreadsheet.getColumns().slice(range._range.left + 1)
+    const cells: Tabulator.CellComponent[][] = rows.map((row) => {
+      const arr = [];
+      row.getCells().forEach((cell) => {
+        if (columns.includes(cell.column)) {
+          arr.push(cell.getComponent());
+        }
+      });
+      return arr;
+    });
+
+    parsedText.data.forEach((row: string[], rowIdx) => {
+      row.forEach((text, colIdx) => {
+        const cell = cells[rowIdx]?.[colIdx]
+        if (!cell) return;
+        setCellValue(cell, text);
+      })
+    })
+  }
+}
+
+export function setCellValue(cell: Tabulator.CellComponent, value: string) {
+  /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+  // @ts-ignore
+  const editableFunc = cell.getColumn().getDefinition().__spreadsheet_editable
```

Comment:
yuck, can we add an API for this on the tabulator side?

Reply:
This is the part where Oli wanted to discuss lol. Column definitions are being mutated by the spreadsheet module and we shouldn't do that. I think he said there were some use cases to fix it so I'll take a look.